Add PHP7 compatible versions for the Null/True/False constraints as they are reserved words in PHP7

stefanruijsenaars opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | PHP7 compatibility |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | N/A |
| Fixed tickets | N/A - helps towards https://github.com/symfony/symfony/issues/14086 |
| License | MIT |

Null, True and False are reserved words in PHP7:

https://wiki.php.net/rfc/reserve_more_types_in_php_7

stefanruijsenaars committed:
Fixed compatibility with PHP7 and up by introducing new constraints (IsNull, IsTrue, IsFalse) and related validators (IsNullValidator, IsTrueValidator, IsFalseValidator)

stefanruijsenaars commented:
We can actually provide the deprecated classes through a class_alias...

Tobion commented:
You missed to rename the file `NullValidatorTest.php`

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Null.php:
```diff
@@ -20,8 +20,7 @@
-{
-    public $message = 'This value should be null.';
-}
+class_alias('NullConstraint', 'Null');
```
The full namespace must be provided

nicolas-grekas replied:
In order to keep type hint working (`... instanceof Null`), this class_alias should be moved at the end of NullConstraint.php, and this very line should be replaced by a class_exists('Symfony\Component\Validator\Constraints\NullConstraints')

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/NullValidator.php:
```diff
@@ -18,6 +18,8 @@
  *
  * @api
+ *
+ * @deprecated
```
Is it really required to deprecate this one also? Deprecations always have a cost for end users, the less the better to me.

stefanruijsenaars replied:
It's actually not.. Should we still add NullConstraintValidator for consistency though?

stefanruijsenaars replied:
Per @stof "The NullValidator needs to be deprecated too"

@stof are we deprecating this or not?

stof replied:
given you are replacing it with the IsNullValidator, the old class must be deprecated yes (and extend the new one to avoid duplicating the logic)

@nicolas-grekas the constraint validators are generally not instantiated by users anyway (especially when there is no dependency like here). The class kept for BC is only for hardcore users wanting to do weird things with the validator

nicolas-grekas replied:
`@deprecated since version 2.3.28, to be removed in 3.0. Use IsNullValidator instead.`

nicolas-grekas commented:
You should do the False and True constraints in the same PR I believe

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+New version
```
2.3.28

nicolas-grekas commented:
Do we really want to do that on 2.3? Deprecations are usually forbidden on 2.3.
I'd personally be fine with having PHP7 being fully supported with 2.7 only.

GromNaN commented:
About the new class name. What about IsNull instead ? That way we can write Assert\IsNull.

Tobion commented:
We don't deprecate things in maintenance branches. This must be done in 2.7

fabpot commented:
The only reason to add those deprecation in 2.3 would be to allow 2.3 to be PHP 7 compatible. Is that what we want?

WouterJ commented:
What about `Nil` and `NotNil`?

stefanruijsenaars commented:
If 2.3 is still supported by the time PHP 7 comes out it would make sense to put this in 2.3, per https://github.com/symfony/symfony/issues/14086 we're quite close to being compatible so if we drop PHP 7 support it shouldn't be just because of deprecation of a few validation constraints :)

Does anyone have any further ideas about new names for the following constraints, I've already seen the following:
NotNull: NotNull (*as it's not a reserved word) / NotNil / IsNotNull / NotNullConstraint
Null: Nil / IsNull  / NullConstraint
True: IsTrue / TrueConstraint
False: IsFalse / FalseConstraint

sstok commented:
I would vote for isNull as its descriptive and consistent with isTrue.

stefanruijsenaars commented:
I will add the new True and False constraints to this PR as well once this Null constraint has been properly reviewed and there's a consensus on the new naming.

Just FYI, this is currently causing a test failure on Drupal 8/PHP 7 as it uses the Null constraint in a test.

dosten commented:
@WouterJ :+1: Also `Truthy` and `Falsy`. What do you think?

stefanruijsenaars commented:
Thanks all, just to be clear, what are we renaming these constraints to?

2015-04-06 15:53 GMT+02:00 Diego Saint Esteben <notifications@github.com>:

> @WouterJ <https://github.com/WouterJ> [image: :+1:] Also Truthy and Falsy.
> What do you think?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-90071266>.
>

stof commented on src/Symfony/Component/Validator/Constraints/NullConstraint.php:
```diff
@@ -0,0 +1,29 @@
+    public $message = 'This value should be null.';
+}
+
+class_alias('Symfony\Component\Validator\Constraints\NullConstraint', 'Symfony\Component\Validator\Constraints\Null');
```
this line should be skipped on PHP7, otherwise the issue is not solved

stof commented:
The NullValidator needs to be deprecated too

stefanruijsenaars commented:
Does this seem ready? What still needs to happen here?

xabbuh commented on src/Symfony/Component/Validator/Constraints/IsNull.php:
```diff
@@ -0,0 +1,32 @@
+    public $message = 'This value should be null.';
+}
+
+if (version_compare(PHP_VERSION, '7.0.0-dev') < 0) {
```
Can't we use `PHP_VERSION_ID < 70000` instead?

jakzal replied:
We should. For consistency.

stof commented on src/Symfony/Component/Validator/Constraints/IsNull.php:
```diff
@@ -0,0 +1,32 @@
+
+if (PHP_VERSION_ID < 70000) {
+  // Null is a reserved word in PHP7.
+  class_alias('Symfony\Component\Validator\Constraints\IsNull', 'Symfony\Component\Validator\Constraints\Null');
```
does it need to be in this file ? couldn't it be in the ``Null.php`` file instead ?

stof commented:
@stefanruijsenaars the comment from @nicolas-grekas in https://github.com/symfony/symfony/pull/14228#issuecomment-89972754 has not been taken into account

stof commented:
@stefanruijsenaars note that your commits are not associated to your github account (the email used when committing is not known by github). You might want to change this

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Null.php:
```diff
@@ -20,8 +18,7 @@
-{
-    public $message = 'This value should be null.';
-}
+class_exists('Symfony\Component\Validator\Constraints\IsNull');
```
I agree with @stof above: this should be replaced by `class_alias`

webmozart replied:
I agree as well. @stefanruijsenaars could you do this change?

Tobion commented:
The `trigger_error` calls must be added.

Tobion commented:
And the same approach should be done for `False` and `True`.

stefanruijsenaars commented:
As soon as I have confirmation that this looks good as far as Null is concerned I can do the same for False and True...

nicolas-grekas commented:
Is it possible that someone has type hinted `Constraints\Null` or `Constraints\NullValidator` (same question for False/True?)
Because the new classes do not pass the deprecated hints, which is a BC break.
To make it pass, we much make `IsNullValidator` extends the deprecated `NullValidator`.
We must also alias `Null` when loading `IsNull`, and load IsNull from Null.php using class_exists (I know, I said the opposite yesterday, sorry for the mistake...)

nicolas-grekas commented:
Btw, there is no need for checking PHP_VERSION_ID. This should be done whatever the version of PHP. You should also use __NAMESPACE__ instead of the literal namespace in class_exists

stof commented:
@nicolas-grekas I don't see any reason for someone to typehint the NullValidator. constraint validators are used inside the validator, not inside your own code.

And the version check needs to be done: we cannot create the ``Null`` alias on PHP 7 given it is a forbidden class name (which is precisely the reason for this renaming)

nicolas-grekas commented:
@stof OK for the PHP7 check (http://3v4l.org/t8u78)
For the type hint, that does not mean we should do it properly.

stof commented:
@nicolas-grekas when renaming a class, the best way to make it pass typehints is actually with a class_alias. It is the only way to make the typehint pass both ways (if you make the new class extend the old one, people using the deprecated class because they rely on the BC layer cannot call code typehinting the new class, which means you have to typehint the deprecated class everywhere, which is a pain)

nicolas-grekas commented:
@stof class_alias is terrible idea: it breaks the DebugClassLoader (or an equivalent strategy) from detecting deprecated use of inheritance.
In this case, we have no choice, because for are forbidden from inheriting from the deprecated class (Null).
Except in this specific case, having the new inherit from the deprecated is exactly the same type hint wise: new and old pass the old hints when the BC layer is in place. When the BC layer is removed, so is the class_alias, and hint break the same way in one case or another.

nicolas-grekas commented:
Oh, and we do not want the deprecated class to pass the new hints. If someone creates a new interface with a new hint, than you must give it a new object, not the old one.

stof commented:
@nicolas-grekas but this means that as long as we keep a BC layer in place, our code is forced to keep typehinting the deprecated class to have a BC layer working (otherwise, we would have to ensure that none of the new features can be called for cases where the user instantiated the BC layer, which can be hard, because it means we cannot rely on the new codebase to implement BC layers).

If you look at existing BC layers, most places are extending classes in a way allowing BC classes to be passed to the codebase using the new typehint (allowing to update the code to stop relying on deprecated classes)

nicolas-grekas commented:
@stof our code is forced to keep the deprecated type hint. This is coherent with the BC policy that forbids changing it. For userland code, once a deprecation is added, all the symfony code base must be updated to use only the new objects so the new type hint can be used very fast.

> most places are extending classes in a way allowing BC classes to be passed to the codebase using the new typehint

I'm not sure of this. See #14223 for an example of what I'm thinking about. See also:
```sh
grep 'extends Legacy' -RI
Component/Validator/Context/ExecutionContextInterface.php:interface ExecutionContextInterface extends LegacyExecutionContextInterface
Component/Validator/Mapping/MetadataInterface.php:interface MetadataInterface extends LegacyMetadataInterface
Component/Validator/Mapping/Factory/MetadataFactoryInterface.php:interface MetadataFactoryInterface extends LegacyMetadataFactoryInterface
Component/Debug/Exception/FlattenException.php:class FlattenException extends LegacyFlattenException
Component/Debug/Exception/FatalErrorException.php:class FatalErrorException extends LegacyFatalErrorException
```

nicolas-grekas commented:
Not to say that class_alias() breaks static analyses also.

stof commented:
@nicolas-grekas static analysis tools could discover calls to ``class_alias`` (unless you use variables coming from elsewhere in it, the static analysis tool can now which alias is beign defined).

Regarding the inheritance order, you found 5 occurrences. Now, count the number of deprecated classes we have doing things the opposite way. There are much more than 5.

nicolas-grekas commented:
The other occurrences are mistakes that have no consequences because they are not used as type hints.

stefanruijsenaars commented:
Could anyone summarize what changes we still need or is this still under
discussion?

2015-04-23 16:32 GMT+02:00 Nicolas Grekas <notifications@github.com>:

> The other occurrences are mistakes that have no consequences because they
> are not used as type hints.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-95606212>.
>

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Null.php:
```diff
@@ -20,8 +18,11 @@
+
+if (PHP_VERSION_ID < 70000) {
+  // Null is a reserved word in PHP7.
+  class_alias('Symfony\Component\Validator\Constraints\IsNull', 'Symfony\Component\Validator\Constraints\Null');
```
`class_alias(__NAMESPACE__.'\IsNull', __NAMESPACE__.'Null');`?

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Null.php:
```diff
@@ -20,8 +18,11 @@
  *
  * @api
+ *
+ * @deprecated
```
`@deprecated since version 2.3.28, to be removed in 3.0. Use IsNull instead.`

nicolas-grekas commented:
@stefanruijsenaars I just added a few minor comments, otherwise I'm :+1:
 
The discussion was a bit more general with @stof and I'm fine with not applying my rant here: if nobody did type hint any of the deprecated classes, which looks likely, then the current patch won't break anyones code. If that holds false, we could easily fix it later. Currently, the patch is simpler, thus pragmatic.

For triggering notices on top of this deprecation, #14472 will do the job.

stefanruijsenaars commented:
OK cool! I have committed your changes and added True and False. Let's give
this a final review and get this in :)

2015-04-26 10:17 GMT+02:00 Nicolas Grekas <notifications@github.com>:

> @stefanruijsenaars <https://github.com/stefanruijsenaars> I just added a
> few minor comments, otherwise I'm [image: :+1:]
>
> The discussion was a bit more general with @stof <https://github.com/stof>
> and I'm fine with not applying my rant here: if nobody did type hint any of
> the deprecated classes, which looks likely, then the current patch won't
> break anyones code. If that holds false, we could easily fix it later.
> Currently, the patch is simpler, thus pragmatic.
>
> For triggering notices on top of this deprecation, #14472
> <https://github.com/symfony/symfony/pull/14472> will do the job.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-96342172>.
>

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/FalseValidator.php:
```diff
@@ -11,27 +11,14 @@
  * @api
+ *
+ * @deprecated Deprecated since version 2.3.28, to be removed in 3.0. Use
+ *             instead.
```
missing: IsFalseValidator

stefanruijsenaars replied:
maybe it was missing in an older commit, I think it's in there now right?

nicolas-grekas replied:
I can't see it even after refreshing the page

stefanruijsenaars replied:
Added now, thanks :)

nicolas-grekas commented:
Looks like a test is broken. Could you please squash the commits also?

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/False.php:
```diff
@@ -11,17 +11,19 @@
-use Symfony\Component\Validator\Constraint;
-
-/**
+/*
```
Double `*`

nicolas-grekas commented on src/Symfony/Component/Validator/Constraints/Null.php:
```diff
@@ -11,17 +11,19 @@
-use Symfony\Component\Validator\Constraint;
-
-/**
+/*
```
Double `*`

stefanruijsenaars replied:
How do we keep the double * from triggering a coding standards warning? (as it's above an empty line and I don't think we can put the docblock right above an if statement)

nicolas-grekas replied:
Hum, that's true. That makes me wonder: how about replacing the class_alias with a proper class declaration?
`class Null extends IsNull {}`
This file will never be loaded under PHP7, so that will not trigger any error!
The good thing is that we'll be able to keep a proper docbloc comment, for the benefit of IDEs like phpStorm.

stefanruijsenaars commented:
Could anyone clarify where the error in the checks and the failed build comes from? I don't see any test fails anymore...

nicolas-grekas commented:
@stefanruijsenaars can you please rebase and squash?

stefanruijsenaars commented:
Will do, but there still seem to be a couple of unexplained test failures.
Are those just flukes?

2015-05-04 18:04 GMT+02:00 Nicolas Grekas <notifications@github.com>:

> @stefanruijsenaars <https://github.com/stefanruijsenaars> can you please
> rebase and squash?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-98766645>.
>

nicolas-grekas commented:
Let's see if they still occur after the rebase

stefanruijsenaars commented:
Done

2015-05-04 18:16 GMT+02:00 Nicolas Grekas <notifications@github.com>:

> Let's see if they still occur after the rebase
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-98770137>.
>

nicolas-grekas commented:
:+1:
This PR allows going a few tests farther on PHP7. Still, we can see a fatal error for the Validator component. It seems unrelated but I don't understand it:
https://travis-ci.org/symfony/symfony/jobs/61229099#L560
A bug in PHP7?
Anyway, good to merge for me.

stof commented:
Too bad that XDebug is not yet available for PHP 7. It would be much easier to debug this with a stack trace for the fatal error

stof commented on src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php:
```diff
@@ -155,8 +155,8 @@ public function guessTypeForConstraint(Constraint $constraint)
-            case 'Symfony\Component\Validator\Constraints\True':
-            case 'Symfony\Component\Validator\Constraints\False':
+            case 'Symfony\Component\Validator\Constraints\IsTrue':
+            case 'Symfony\Component\Validator\Constraints\IsFalse':
```
the deprecated names should be kept here (this is the issue when using a ``switch(get_class()``: it does not support inheritance)

stefanruijsenaars replied:
Done

2015-05-07 15:00 GMT+02:00 Christophe Coevoet <notifications@github.com>:

> In src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php
> <https://github.com/symfony/symfony/pull/14228#discussion_r29848859>:
>
> > @@ -155,8 +155,8 @@ public function guessTypeForConstraint(Constraint $constraint)
> >              case 'Symfony\Component\Validator\Constraints\Count':
> >                  return new TypeGuess('collection', array(), Guess::LOW_CONFIDENCE);
> >
> > -            case 'Symfony\Component\Validator\Constraints\True':
> > -            case 'Symfony\Component\Validator\Constraints\False':
> > +            case 'Symfony\Component\Validator\Constraints\IsTrue':
> > +            case 'Symfony\Component\Validator\Constraints\IsFalse':
>
> the deprecated names should be kept here (this is the issue when using a
> switch(get_class(): it does not support inheritance)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228/files#r29848859>.
>

stof commented on src/Symfony/Component/Form/Extension/Validator/ValidatorTypeGuesser.php:
```diff
@@ -173,7 +175,7 @@ public function guessRequiredForConstraint(Constraint $constraint)
         switch (get_class($constraint)) {
             case 'Symfony\Component\Validator\Constraints\NotNull':
             case 'Symfony\Component\Validator\Constraints\NotBlank':
-            case 'Symfony\Component\Validator\Constraints\True':
```
you forgot to keep it here

stof commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -64,7 +64,7 @@ public function guessRequiredProvider()
             array(new NotNull(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
             array(new NotBlank(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
-            array(new True(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
+            array(new IsTrue(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
```
you should keep a legacy test (skipped on PHP 7) using the old constraint

stefanruijsenaars replied:
Thanks, committed

2015-05-07 15:12 GMT+02:00 Christophe Coevoet <notifications@github.com>:

> In
> src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php
> <https://github.com/symfony/symfony/pull/14228#discussion_r29849742>:
>
> > @@ -64,7 +64,7 @@ public function guessRequiredProvider()
> >          return array(
> >              array(new NotNull(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
> >              array(new NotBlank(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
> > -            array(new True(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
> > +            array(new IsTrue(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
>
> you should keep a legacy test (skipped on PHP 7) using the old constraint
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228/files#r29849742>.
>

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -64,7 +64,10 @@ public function guessRequiredProvider()
             array(new NotNull(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
             array(new NotBlank(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
-            array(new True(), new ValueGuess(true, Guess::HIGH_CONFIDENCE)),
+            if (PHP_VERSION_ID < 70000) {
```
Parse error here. Le legacy test should be split in a dedicated `public function testLegacyGuessRequired()` test method, with the `@group legacy` annotation.

aitboudad commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -19,6 +19,7 @@
 use Symfony\Component\Validator\Constraints\NotNull;
 use Symfony\Component\Validator\Constraints\Range;
+use Symfony\Component\Validator\Constraints\IsTrue;
 use Symfony\Component\Validator\Constraints\True;
```
can you try with `use ..\True as LegacyTrue;` or is there another way to avoid Fatal error ?
`PHP Fatal error:  Cannot use Symfony\Component\Validator\Constraints\True as True because 'True' is a special class name in `

stefanruijsenaars replied:
Yes, let's try that. Otherwise we can wrap the use statement in an if block
and test for the PHP version.

2015-05-10 13:28 GMT+02:00 Abdellatif Ait boudad <notifications@github.com>:

> In
> src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php
> <https://github.com/symfony/symfony/pull/14228#discussion_r30001164>:
>
> > @@ -19,6 +19,7 @@
> >  use Symfony\Component\Validator\Constraints\NotBlank;
> >  use Symfony\Component\Validator\Constraints\NotNull;
> >  use Symfony\Component\Validator\Constraints\Range;
> > +use Symfony\Component\Validator\Constraints\IsTrue;
> >  use Symfony\Component\Validator\Constraints\True;
>
> can you try with use ..\True as LegacyTrue; or is there another way to
> avoid Fatal error ?
> PHP Fatal error: Cannot use Symfony\Component\Validator\Constraints\True
> as True because 'True' is a special class name in
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228/files#r30001164>.
>

aitboudad commented:
:+1:

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -19,7 +19,8 @@
 use Symfony\Component\Validator\Constraints\Range;
-use Symfony\Component\Validator\Constraints\True;
+use Symfony\Component\Validator\Constraints\IsTrue;
+use Symfony\Component\Validator\Constraints\True as LegacyTrue;
```
It won't work. We must use a string is the method

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -84,6 +85,25 @@ public function testGuessRequired($constraint, $guess)
+    }
+
+    /**
+     * @dataProvider legacyGuessRequiredProvider
```
We don't need the data provider

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -84,6 +87,16 @@ public function testGuessRequired($constraint, $guess)
+    public function testLegacyGuessRequired($constraint, $guess)
+    {
+      if (PHP_VERSION_ID < 70000) {
+        $this->testGuessRequired(new True(), new ValueGuess(true, Guess::HIGH_CONFIDENCE));
```
new $true, where $true = 'Symfony\…\True'

nicolas-grekas replied:
Also would be better to move the test outside of this if and call markTestSkipped here

stefanruijsenaars replied:
So the test just underneath the if block and markTestSkipped inside?

xabbuh commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -84,6 +84,18 @@ public function testGuessRequired($constraint, $guess)
+    public function testLegacyGuessRequired()
+    {
+      if (PHP_VERSION_ID >= 70000) {
+        $this->markTestSkipped('PHP is version 7 or higher.');
```
Imho the message should provide the reason why this is kipped (something like "Cannot use a class called True on PHP 7 or higher").

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -84,6 +84,18 @@ public function testGuessRequired($constraint, $guess)
+    public function testLegacyGuessRequired()
+    {
+      if (PHP_VERSION_ID >= 70000) {
+        $this->markTestSkipped('Cannot use a class called True on PHP 7 or higher.');
```
two indenting spaces missing

stefanruijsenaars commented:
The checks seems stuck, can we re-queue this somehow?

nicolas-grekas commented on src/Symfony/Component/Form/Tests/Extension/Validator/ValidatorTypeGuesserTest.php:
```diff
@@ -84,6 +84,18 @@ public function testGuessRequired($constraint, $guess)
+     */
+    public function testLegacyGuessRequired()
+    {
+      if (PHP_VERSION_ID >= 70000) {
```
two indentation spaces missing for the whole method

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -1,6 +1,16 @@
 CHANGELOG
 =========
 
+2.3.28
```
2.3.29

Berdir commented:
I wasn't able to reproduce the recursion that travis reports. The validator tests are running through for me on php7 (PHP 7.0.0-dev (cli) (built: May  6 2015 00:41:14)).

However, I have a lot of test fails, all because the string representation of a DateTime object is somehow different, it expects 'Jan 1, 2000, 12:00 AM' but actually gets 'Jan 1, 2000 12:00 AM' (note the missing , after the year). But that seems like a different problem and could be environment specific.

nicolas-grekas commented:
ping @symfony/deciders

Tobion commented on src/Symfony/Component/Validator/Constraints/NullValidator.php:
```diff
@@ -11,25 +11,12 @@
  *
  * @api
+ *
+ * @deprecated Deprecated since version 2.3.28, to be removed in 3.0. Use
```
We do not add deprecations in patch version. So this comment should should only be added in 2.7.

Tobion replied:
together with trigger_errors

stefanruijsenaars replied:
So trigger_errors in 2.7 only?

This discussion had appeared earlier in this thread as well, I think back then it was decided to make an exception? Though @fabpot did mention the deprecations would only be necessary if we wanted to make the branch PHP7 compatible.

LionsAd replied:
I just talked with @fabpot in person and yes the deprecation and trigger_error is only in 2.7.

2.3 will be PHP 7 compatible for Symfony for out-of-the-box, but we won't prevent others from still using the old constraints, etc. (and such make it PHP 7 incompatible)

That way 2.3 is compatible out-of-the-box for Drupal + PHP 7, but you could still shoot yourself in the foot with a custom extension.

In 2.7 then it will start with deprecation notices.

LionsAd commented:
So we probably need a 2.7 PR as well?

nicolas-grekas commented:
@stefanruijsenaars can you please remove any comment telling about deprecation (in the patch but also in the PR title/description)?
We are going to consider this as a pure bug fix on 2.3, and we will make and other PR on 2.7 to add the deprecation back after the merge.

stefanruijsenaars commented:
Done

2015-05-13 10:33 GMT+02:00 Nicolas Grekas <notifications@github.com>:

> @stefanruijsenaars <https://github.com/stefanruijsenaars> can you please
> remove any comment telling about deprecation (in the patch but also in the
> PR title/description)?
> We are going to consider this as a pure bug fix on 2.3, and we will make
> and other PR on 2.7 to add the deprecation back after the merge.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#issuecomment-101572918>.
>

nicolas-grekas commented on src/Symfony/Component/Validator/CHANGELOG.md:
```diff
@@ -1,6 +1,12 @@
+------
+
+ * added PHP7-compatible versions for the constraints Null, True and False (IsNull, IsTrue, IsFalse)
+ * added PHP7-compatible versions for the constraint validators NullValidator, TrueValidator and FalseValidator (IsNullValidator, IsTrueValidator, IsFalseValidator)
```
I propose a single line:
- fixed compatibility with PHP7 and up by introducing new constraints (IsNull, IsTrue, IsFalse) and related validators (IsNullValidator, IsTrueValidator, IsFalseValidator)

nicolas-grekas commented:
Looking at the tests, you need to patch one more file:
https://github.com/stefanruijsenaars/symfony/blob/php7-nullconstraint/src/Symfony/Component/Form/composer.json#L28
should be:
`        "symfony/validator": "~2.3.0,>=2.3.29",`

xabbuh commented:
Why not using `~2.3.29` instead then?

nicolas-grekas commented:
@xabbuh indeed

Tobion commented:
:+1:

stof commented:
:+1:

stof commented:
Thanks for your work on this @stefanruijsenaars.

stefanruijsenaars commented:
Do we have a 2.7 PR for this yet?

2015-05-15 11:26 GMT+02:00 Christophe Coevoet <notifications@github.com>:

> Merged #14228 <https://github.com/symfony/symfony/pull/14228>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/symfony/symfony/pull/14228#event-305678429>.
>

nicolas-grekas commented:
@stefanosala it is already in 2.7, check the git log

nicolas-grekas commented:
@stefanruijsenaars it is already in 2.7, check the git log

fgleyze commented:
yes
